payjoin/src/receive/multi_party/mod.rs

```diff
@@ -0,0 +1,268 @@
+    ) -> Result<OutputsUnknown, Error> {
+        let inner = self.v1.check_no_inputs_seen_before(is_seen)?;
+        Ok(OutputsUnknown { v1: inner, sender_contexts: self.sender_contexts })
+    }
+}
+
+pub struct OutputsUnknown {
+    v1: v1::OutputsUnknown,
+    sender_contexts: Vec<SessionContext>,
+}
+
+impl OutputsUnknown {
+    pub fn identify_receiver_outputs(
+        self,
+        is_receiver_output: impl Fn(&bitcoin::Script) -> Result<bool, Error>,
+    ) -> Result<WantsOutputs, Error> {
+        let inner = self.v1.identify_receiver_outputs(is_receiver_output)?;
+        Ok(WantsOutputs { v1: inner, sender_contexts: self.sender_contexts })
+    }
+}
+
+pub struct WantsOutputs {
+    v1: v1::WantsOutputs,
+    sender_contexts: Vec<SessionContext>,
+}
+
+impl WantsOutputs {
+    pub fn commit_outputs(self) -> WantsInputs {
+        let inner = self.v1.commit_outputs();
+        WantsInputs { v1: inner, sender_contexts: self.sender_contexts }
+    }
+}
+
+pub struct WantsInputs {
+    v1: v1::WantsInputs,
+    sender_contexts: Vec<SessionContext>,
+}
+
+impl WantsInputs {
+    /// Add the provided list of inputs to the transaction.
+    /// Any excess input amount is added to the change_vout output indicated previously.
+    pub fn contribute_inputs(
+        self,
+        inputs: impl IntoIterator<Item = InputPair>,
+    ) -> Result<WantsInputs, InputContributionError> {
+        let inner = self.v1.contribute_inputs(inputs)?;
+        Ok(WantsInputs { v1: inner, sender_contexts: self.sender_contexts })
+    }
+
+    /// Proceed to the proposal finalization step.
+    /// Inputs cannot be modified after this function is called.
+    pub fn commit_inputs(self) -> ProvisionalProposal {
+        let inner = self.v1.commit_inputs();
+        ProvisionalProposal { v1: inner, sender_contexts: self.sender_contexts }
+    }
+}
+
+pub struct ProvisionalProposal {
+    v1: v1::ProvisionalProposal,
+    sender_contexts: Vec<SessionContext>,
+}
+
+impl ProvisionalProposal {
+    pub fn finalize_proposal(
+        self,
+        wallet_process_psbt: impl Fn(&Psbt) -> Result<Psbt, Error>,
+        min_feerate_sat_per_vb: Option<FeeRate>,
+        max_feerate_sat_per_vb: FeeRate,
+    ) -> Result<PayjoinProposal, Error> {
+        let inner = self.v1.finalize_proposal(
+            wallet_process_psbt,
+            min_feerate_sat_per_vb,
+            Some(max_feerate_sat_per_vb),
+        )?;
+        Ok(PayjoinProposal { v1: inner, sender_contexts: self.sender_contexts })
+    }
+}
+
+pub struct PayjoinProposal {
+    v1: v1::PayjoinProposal,
+    sender_contexts: Vec<SessionContext>,
+}
+
+impl PayjoinProposal {
+    pub fn sender_iter(&self) -> impl Iterator<Item = v2::PayjoinProposal> {
+        self.sender_contexts
+            .iter()
+            .map(|ctx| v2::PayjoinProposal::new(self.v1.clone(), ctx.clone()))
+            .collect::<Vec<_>>()
+            .into_iter()
+    }
```

Comment:
Initially, I felt like it might make sense to directly `extract_v2_reqs() -> Vec<(req, ctx)>` instead of returning the proposals which requires cloning, but I do understand that each proposal may need to poll a different number of times.

I do have some pause cloning the `ctx` because it's supposed to be mutated on each request so that each request has a new nonce. Not sure if cloning the session context is going to allow v2::PayjoinProposal to do this or if, upon e.g. pause and resume, the receiver would be reuse key material. 🤔